query-algebrizer/src/clauses/predicate.rs

```diff
@@ -81,13 +94,181 @@ impl ConjoiningClauses {
+        // // TODO: should this just use `resolve_argument`?  Should it add a new `resolve_*` function?
+        // let search = match args.next().unwrap() {
+        //     FnArg::Variable(var) => {
+        //         self.column_bindings
+        //             .get(&var)
+        //             .and_then(|cols| cols.first().map(|col| QueryValue::Column(col.clone())))
+        //             .ok_or_else(|| Error::from_kind(ErrorKind::UnboundVariable(var.name())))?
+        //     },
+        //     FnArg::Constant(NonIntegerConstant::Text(s)) => {
+        //         QueryValue::TypedValue(TypedValue::typed_string(s.as_str()))
+        //     },
+        //     _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "string".into(), 2)),
+        // };
+
+        // // TODO: should we build the FQA in ::Matches, preventing nonsense like matching on ::Rowid?
+        // let constraint = ColumnConstraint::Matches(QualifiedAlias(fulltext_values_alias.clone(), Column::Fulltext(FulltextColumn::Text)), search);
+        // self.wheres.add_intersection(constraint);
+
+        let mut values: Vec<Vec<TypedValue>> = Default::default();
+
+        let mut projected: Vec<Variable> = Default::default();
+
+        // TODO: binding-specific error messages.
+        match where_fn.binding {
+            Binding::BindScalar(var) => {
+                // if bindings.len() > 4 {
+                //     bail!(ErrorKind::InvalidNumberOfArguments(where_fn.operator.clone(), bindings.len(), 4));
+                // }
+                // bindings.into_iter()
+
+                if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                    // That's a paddlin'!
+                    bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "illegal bound variable".into(), 999))
+                }
+
+                let constant: TypedValue = match args.next().unwrap() {
+                    FnArg::EntidOrInteger(x) => {
+                        TypedValue::Long(x)
+                    },
+                    FnArg::Ident(x) => {
+                        TypedValue::Keyword(Rc::new(x))
+                    },
+                    FnArg::Constant(NonIntegerConstant::Boolean(x)) => TypedValue::Boolean(x),
+                    FnArg::Constant(NonIntegerConstant::BigInteger(x)) => unimplemented!(),
+                    FnArg::Constant(NonIntegerConstant::Float(x)) => TypedValue::Double(x),
+                    FnArg::Constant(NonIntegerConstant::Text(x)) => TypedValue::String(x),
+                    FnArg::Variable(_) | FnArg::SrcVar(_) => panic!("bad things happened"),
+                    FnArg::Collection(_) => panic!("more bad things happened"),
+                };
+
+                projected.push(var.clone());
+                self.constrain_var_to_type(var.clone(), constant.value_type());
+
+                values.push(vec![constant]); // 1x1.
+
+                // // TODO: this optimization.
```

Comment:
There're two kinds of optimization to do here, no?

1. This one: if we haven't seen this variable yet, and we are binding it to a scalar, then just stuff it into `value_bindings` and pretend we never saw the `ground`. In that respect `ground` becomes no different to `:in`.
2. If the variable has already been bound earlier in the query, `ground` _becomes a constraint_:

```
[:find ?x 
 :where [?x :foo/name ?y]
 [(ground "foo") [?y .]]]
```

is

```
SELECT datoms00.e FROM datoms datoms00
WHERE datoms00.a = 99 AND datoms00.v = 'foo'
```

I'm not sure whether further substitution is valuable after that — SQLite might be able to do index optimizations with a column that we can't do by pushing in a value.


While I'm typing, we also should handle this:

```
[:find … :in ?x :where [(ground "foo") ?x]
```

— it doesn't really make sense to ground a constant to an input variable. (Not sure if you've already addressed this.)